jetty-http/src/main/java/org/eclipse/jetty/http/HttpMethod.java

```diff
@@ -26,162 +26,179 @@
+     * @return true if the method is idempotent.
+     */
+    public boolean isIdempotent()
     {
-        for (HttpMethod method : HttpMethod.values())
-        {
-            CACHE.put(method.toString(), method);
-        }
+        return _type.ordinal() >= Type.IDEMPOTENT.ordinal();
     }
 
-    private final ByteBuffer _buffer;
-    private final byte[] _bytes;
+    public ByteBuffer asBuffer()
+    {
+        return _buffer.asReadOnlyBuffer();
+    }
 
-    HttpMethod()
+    public String asString()
     {
-        _bytes = StringUtil.getBytes(toString());
-        _buffer = ByteBuffer.wrap(_bytes);
+        return _method;
     }
 
-    public byte[] getBytes()
+    public String toString()
     {
-        return _bytes;
+        return _method;
     }
 
-    public boolean is(String s)
+    public static final Trie<HttpMethod> INSENSITIVE_CACHE = new ArrayTrie<>(252);
+    public static final Trie<HttpMethod> CACHE = new ArrayTernaryTrie<>(false, 300);
+    public static final Trie<HttpMethod> LOOK_AHEAD = new ArrayTernaryTrie<>(false, 330);
+    static
     {
-        return toString().equalsIgnoreCase(s);
+        for (HttpMethod method : HttpMethod.values())
+        {
+            if (!INSENSITIVE_CACHE.put(method.asString(), method))
+                throw new IllegalStateException("INSENSITIVE_CACHE too small: " + method);
+
+            if (!CACHE.put(method.asString(), method))
+                throw new IllegalStateException("CACHE too small: " + method);
+
+            if (!LOOK_AHEAD.put(method.asString() + ' ', method))
+                throw new IllegalStateException("LOOK_AHEAD too small: " + method);
+        }
     }
 
-    public ByteBuffer asBuffer()
+    /**
+     * Optimized lookup to find a method name and trailing space in a byte array.
+     *
+     * @param bytes Array containing ISO-8859-1 characters
+     * @param position The first valid index
+     * @param limit The first non valid index
+     * @return An HttpMethod if a match or null if no easy match.
+     */
+    public static HttpMethod lookAheadGet(byte[] bytes, final int position, int limit)
     {
-        return _buffer.asReadOnlyBuffer();
+        int len = limit - position;
+        if (limit > 3)
+        {
+            // Short cut for GET
+            if (bytes[position] == 'G' && bytes[position + 1] == 'E' && bytes[position + 2] == 'T' && bytes[position + 3] == ' ')
+                return GET;
+            // Otherwise lookup in the Trie
+            return LOOK_AHEAD.getBest(bytes, position, len);
+        }
+        return null;
     }
 
-    public String asString()
+    /**
+     * Optimized lookup to find a method name and trailing space in a byte array.
+     *
+     * @param buffer buffer containing ISO-8859-1 characters, it is not modified.
+     * @return An HttpMethod if a match or null if no easy match.
+     * @deprecated Not used
```

Comment:
It is still used by `HttpParser`.

Reply:
Now this is an opps- deprecated the wrong one!